daemon/status.go

```diff
@@ -85,90 +81,275 @@ func (d *Daemon) getNodeStatus() *models.ClusterStatus {
+	stale := make(map[string]strfmt.DateTime, len(staleProbes))
+	for probe, startTime := range staleProbes {
+		stale[probe] = strfmt.DateTime(startTime)
 	}
 
-	sr.ContainerRuntime = workloads.Status()
+	d.statusCollectMutex.RLock()
+	defer d.statusCollectMutex.RUnlock()
 
-	sr.Kubernetes = d.getK8sStatus()
+	// TODO(brb) we should do a deep copy instead, as statusResponse contains some references
+	sr := d.statusResponse
+	sr.Stale = stale
 
-	// Note: A final, overriding, check is made in Handle to check the staleness
-	// of this data, and will clobber these messages if set.
-	if sr.Kvstore.State != models.StatusStateOk {
+	switch {
+	case len(sr.Stale) > 0:
+		sr.Cilium = &models.Status{
+			State: models.StatusStateWarning,
+			Msg:   "Stale status data",
+		}
+	case sr.Kvstore != nil && sr.Kvstore.State != models.StatusStateOk:
 		sr.Cilium = &models.Status{
 			State: sr.Kvstore.State,
 			Msg:   "Kvstore service is not ready",
 		}
-	} else if sr.ContainerRuntime.State != models.StatusStateOk {
+	case sr.ContainerRuntime != nil && sr.ContainerRuntime.State != models.StatusStateOk:
 		sr.Cilium = &models.Status{
 			State: sr.ContainerRuntime.State,
 			Msg:   "Container runtime is not ready",
 		}
-	} else if k8s.IsEnabled() && sr.Kubernetes.State != models.StatusStateOk {
+	case k8s.IsEnabled() && sr.Kubernetes != nil && sr.Kubernetes.State != models.StatusStateOk:
 		sr.Cilium = &models.Status{
 			State: sr.Kubernetes.State,
 			Msg:   "Kubernetes service is not ready",
 		}
-	} else {
+	default:
 		sr.Cilium = &models.Status{State: models.StatusStateOk, Msg: "OK"}
 	}
 
-	sr.IPAM = d.DumpIPAM()
-	sr.NodeMonitor = d.nodeMonitor.State()
-
-	sr.Cluster = d.getNodeStatus()
+	return sr
+}
 
-	if d.ciliumHealth != nil {
-		sr.Cluster.CiliumHealth = d.ciliumHealth.GetStatus()
+func (d *Daemon) startStatusCollector() {
+	probes := []status.Probe{
+		{
+			Name: "check-locks",
+			Probe: func(ctx context.Context) (interface{}, error) {
+				// Try to acquire a couple of global locks to have the status API fail
+				// in case of a deadlock on these locks
+				option.Config.ConfigPatchMutex.Lock()
+				option.Config.ConfigPatchMutex.Unlock()
+				return nil, nil
+			},
+			OnStatusUpdate: func(status status.Status) {
+				d.statusCollectMutex.Lock()
+				defer d.statusCollectMutex.Unlock()
+				// FIXME we have no field for the lock status
+			},
+		},
+		{
+			Name: "kvstore",
+			Probe: func(ctx context.Context) (interface{}, error) {
+				return kvstore.Client().Status()
+			},
+			OnStatusUpdate: func(status status.Status) {
+				var msg string
+				state := models.StatusStateOk
+				info, ok := status.Data.(string)
+
+				switch {
+				case ok && status.Err != nil:
+					state = models.StatusStateFailure
+					msg = fmt.Sprintf("Err: %s - %s", status.Err, info)
+				case status.Err != nil:
+					state = models.StatusStateFailure
+					msg = fmt.Sprintf("Err: %s", status.Err)
+				case ok:
+					msg = fmt.Sprintf("%s", info)
+				}
```

Comment:
I don't see the `!ok` case being handled here. Am I missing something? Or can we just ignore situations when `Data` has a different type?

Reply:
I was reluctant to handle the `!ok` case, as it can only happen when someone changes the `"kvstore"`'s `Probe` method in a way that it returns a non-string type, and it would require adding more boilerplate (if we do here, then probably we would need to add to other probes as well).

Reply:
OK, makes sense, thanks for explanation. I'm fine with not handling `!ok` case then.